### PR TITLE
mgr/dashboard: carbonize Cluster > Hosts form

### PR DESCRIPTION
**Pull request by dnyanee1997**
Fixes: https://tracker.ceph.com/issues/68264

Signed-off-by: Dnyaneshwari Talwekar <dtalweka@redbat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html [hunk 1]
```diff
@@ -1,108 +1,112 @@
-<cd-modal [pageURL]="pageURL"
-          [modalRef]="activeModal">
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html:2]: the `pageURL` seems to be missing in the carbon implementation

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html [hunk 2]
```diff
@@ -1,108 +1,112 @@
...
-  <span class="modal-title"
-        i18n>{{ action | titlecase }} {{ resource | upperFirst }}</span>
-
-  <ng-container class="modal-content">
-
-    <div *cdFormLoading="loading">
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html:8]: The `loading` is missing after the changes.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts [hunk 3]
```diff
@@ -46,17 +48,17 @@ export class HostFormComponent extends CdForm implements OnInit {
     private actionLabels: ActionLabelsI18n,
     private hostService: HostService,
     private taskWrapper: TaskWrapperService,
-    public activeModal: NgbActiveModal
+    public activeModal: NgbActiveModal,
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:51]:
```suggestion
```
NgbActiveModal is no longer being used, right?

**Review by Pegonzal — COMMENTED** [review]
Thanks @dnyanee1997 for the PR, just tested it, it is working as expected, but looks lke the helper text is not properly formatted

image

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html [hunk 4]
```diff
@@ -1,108 +1,112 @@
...
+          <input
+            cdsText
+            type="text"
+            placeholder="mon-123"
+            id="hostname"
+            name="hostname"
+            formControlName="hostname"
+            autofocus
+            (keyup)="checkHostNameValue()"
+          />
```
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html]: Following [cds guidelines](https://carbondesignsystem.com/components/text-input/usage/#anatomy) 
helper text should be at the bottom of the input, and doesn't say much else about the styling of that but I believe it should be kept simple, I think the paragraph element should be removed, it takes too much space:
Removed `<p>` tag
image


Removed `<p>` tag and list tags:
image


From which is better that the other I have no idea, both of them work for me

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html [hunk 5]
```diff
@@ -1,108 +1,112 @@
...
+      <!-- Hostname -->
+      <div class="form-item">
+        <cds-text-label
+          label="Hostname"
+          for="hostname"
+          cdRequiredField="Hostname"
+          [invalid]="!hostForm.controls.hostname.valid && hostForm.controls.hostname.dirty"
+          [invalidText]="hostnameError"
+          i18n
+          >Hostname
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html:24]: I can see other places where we are not applying the usual dashboard formatting

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html [hunk 6]
```diff
@@ -1,108 +1,112 @@
...
+                        for="addr"
+                        i18n>Network address
+          <input
+            cdsText
+            type="text"
+            placeholder="192.168.0.1"
+            id="addr"
+            name="addr"
+            formControlName="addr"
+          />
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html:66]: Here as well

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/cluster.module.ts [hunk 7]
```diff
@@ -102,7 +104,10 @@ import { MultiClusterDetailsComponent } from './multi-cluster/multi-cluster-deta
     CheckboxModule,
     GridModule,
     ProgressIndicatorModule,
-    ButtonModule
+    ButtonModule,
+    InputModule,
+    ButtonModule,
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/cluster.module.ts:109]: This is repeated, right?
```suggestion
    ButtonModule,
    InputModule,
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts [hunk 8]
```diff
@@ -15,13 +13,15 @@ import { CdValidators } from '~/app/shared/forms/cd-validators';
...
 import { TaskWrapperService } from '~/app/shared/services/task-wrapper.service';
+import { Location } from '@angular/common';
 
 @Component({
   selector: 'cd-host-form',
   templateUrl: './host-form.component.html',
   styleUrls: ['./host-form.component.scss']
 })
 export class HostFormComponent extends CdForm implements OnInit {
+  open = false;
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:24]:
```suggestion
  open: boolean = false;
```

**Review by Pegonzal — APPROVED** [review]
Thank you @dnyanee1997 for the update, I've tested the PR and it works as expected, the PR mostly looks good, just a couple things I see

There is some linting issue:
```

[lint:scss    ] Browserslist: caniuse-lite is outdated. Please run:
[lint:scss    ]   npx update-browserslist-db@latest
[lint:scss    ]   Why you should do it regularly: https://github.com/browserslist/update-db#readme
[lint:html    ] 
[lint:html    ] [htmllint] found 0 errors out of 244 files
[lint:html    ] All files pass linting
[lint:prettier] src/app/ceph/cluster/cluster.module.ts
[lint:prettier] src/app/ceph/cluster/hosts/host-form/host-form.component.ts
[lint:eslint  ] 
[lint:eslint  ] Linting "ceph-dashboard"...
[lint:eslint  ] 
[lint:eslint  ] All files pass linting.
[lint:eslint  ] 

Try running 'npm run fix' to fix some linting errors. Some errors might need a manual fix.
```

**Comment by dnyanee1997** [comment]
jenkins test make check

**Comment by dnyanee1997** [comment]
Thanks @Pegonzal  for reviewing the PR.

**Comment by dnyanee1997** [comment]
jenkins test dashboard

**Comment by dnyanee1997** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts [hunk 9]
```diff
@@ -94,7 +93,7 @@ export class HostFormComponent extends CdForm implements OnInit {
         validators: [CdValidators.ip()]
       }),
       labels: new UntypedFormControl([]),
-      maintenance: new UntypedFormControl(false)
+      maintenance: new UntypedFormControl()
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:96]: why this change?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts [hunk 10]
```diff
@@ -166,9 +165,13 @@ export class HostFormComponent extends CdForm implements OnInit {
           complete: () => {
             this.pageURL === 'hosts'
               ? this.router.navigate([this.pageURL, { outlets: { modal: null } }])
-              : this.activeModal.close();
+              : this.location.back();
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:168]: as it is already defined as a function below
```suggestion
              : this.closeModal()
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts [hunk 11]
```diff
@@ -153,7 +151,7 @@ export class HostsComponent extends ListWithDetails implements OnDestroy, OnInit
         click: () =>
           this.router.url.includes('/hosts')
             ? this.router.navigate([BASE_URL, { outlets: { modal: [URLVerbs.ADD] } }])
-            : (this.bsModalRef = this.modalService.show(HostFormComponent, {
+            : (this.bsModalRef = this.cdsModalService.show(HostFormComponent, {
                 hideMaintenance: this.hideMaintenance
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:155]: i see that we are passing an input to the modal here but as per carbon we need to Inject that input to the component. I don't see that's happening here? Btw this input might be used in the expand-cluster add host form where we hide the maintenance checkbox. you can cross check that behavior as well.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html [hunk 12]
```diff
@@ -1,108 +1,104 @@
...
+          </span>
+        </ng-template>
+        <cd-help-text>
+          To add multiple hosts at once, you can enter:
+          <ul>
+            <li>a comma-separated list of hostnames <samp>(e.g.: example-01,example-02,example-03)</samp>,</li>
+            <li>a range expression <samp>(e.g.: example-[01-03].ceph)</samp>,</li>
+            <li>a comma separated range expression <samp>(e.g.: example-[01-05].lab.com,example2-[1-4].lab.com,example3-[001-006].lab.com)</samp></li>
+          </ul>
+        </cd-help-text>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html:49]: btw, one more thing as I see the screenshot, maybe this needs to be hidden inside the tooltip (or cd-helper)? I don't know if we ever differentiated between when to use `cd-help-text` and when to use `cd-helper` but this looks like a good example. 

and btw, @dnyanee1997 always try to keep the `i18n` there because its needed for our internationalization purposes

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts (hunk 9 above)
Comment (dnyanee1997) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:96]: Hi @nizamial09 , thanks for brining in. While I was making the changes in the Ui . I found that if we checked the maintenance checkbox, the value we are getting of status: 'maintenance' otherwise its blank and here we are getting "500 internal server error" as status should be boolean one. 


Can you please confirm if status should be **boolean or string**?

Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:96]: it should be a boolean.

**Comment by afreen23** [comment]
Kindly, create a follow-up PR.
I am amazed I did not see them, apologies for that @nizamial09 @dnyanee1997 

Thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts (hunk 9 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:96]: now thatt i checked the code properly, the value of `this.status` is a string. but the form control value is boolean since its a checkbox. the conversion happen here: https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts#L148
